proposals/0292-package-registry-service.md

```diff
@@ -384,160 +744,171 @@ and individual commits may be cryptographically signed by authors.
+      }
+  },
+  "version": 1
+}
 
-```terminal
-$ swift package config set-mirror \
-    --original-url https://github.com/mona/TopSecret \
-    --mirror-url https://<token>:x-oauth-basic@github.com/mona/TopSecret
 ```
 
+This kind of attack can be mitigated on an individual basis
+by adding `.swiftpm/config` to a project's `.gitignore` file.
 The risk could be mitigated for all users
-if Swift Package Manager forbids the use of hardcoded credentials
-in `Package.swift` files.
+if Swift Package Manager included a `.gitignore` file
+in its new project template.
+Code hosting providers can also help minimize this risk
+by [detecting secrets][secret scanning]
+that are committed to public repositories.
+
+Credentials may also be unintentionally disclosed
+by Swift Package Manager or other tools in logging statements.
+Care should be taken to redact the user info component of URLs
+when displaying feedback to the user
+(for example,
+the URL `https://<USERNAME>:<TOKEN>@swift.pkg.github.com`
+is logged as `https://***@swift.pkg.github.com`).
 
 ### Denial of service
 
 An attacker could scrape public code repositories
-for `Package.swift` files that declare dependencies
+for `.swiftpm/config` files that declare one or more custom registries
 and launch a denial-of-service attack
 in an attempt to reduce the availability of those resources.
 
 The likelihood of this attack is generally low
 but could be used in a targeted way
 against resources known to be important or expensive to distribute.
 
-This threat can be mitigated by obfuscating dependency URLs,
-such that they can't be pattern matched from source code.
-
-```swift
-func rot13(_ string: String) -> String {
-    String(string.unicodeScalars.map { unicodeScalar in
-        var value = unicodeScalar.value
-        switch unicodeScalar {
-        case "A"..."M", "a"..."m": value += 13
-        case "N"..."Z", "n"..."z": value -= 13
-        default: break
-        }
-
-        return Character(Unicode.Scalar(value)!)
-    })
-}
-
-dependencies: [
-  .archive(name: "TopSecret",
-           url: rot13("uggcf://tvguho.pbz/zban/GbcFrperg"),
-           //       ^ "https://github.com/mona/TopSecret"
-           checksum: "2c4a4ce92225fb766447c1757abb916e13f68eba0459f1287ee62e4941d89bbf")
-]
-```
-
-> **Important**:
-> Never store credentials in code —
-> _even if they're obfuscated_.
-
 ### Escalation of privilege
 
 Even authentic packages from trusted creators can contain malicious code.
 
 Code analysis tools can help to some degree,
 as can system permissions and other OS-level security features.
-But developers are ultimately the ones responsible
-for the code they ship to users.
+However, developers are ultimately responsible for the code they ship to users.
 
 ## Impact on existing packages
 
 Current packages won't be affected by this change,
-as they'll continue to be able to download dependencies directly through Git.
+as they'll continue to download dependencies directly through Git.
 
```

Comment:
one other threat we can mention is "server driven dos attacks", i.e. a registry returning huge, slow or malicious responses to registry API requests causing damage or resources starvation on the client (SwiftPM, etc). we can address some of them with exponential backoff and circuit breaking on the client side.

Reply:
Hmm, I don't think I've heard the term DOS applied to an attack on the client by the server. Incorporating a backoff, circuit breaker, and/or validation on the client as a hedge against flakey or misbehaving servers seems like good advice for any application. Or do you see a consideration in this regard that's specific to this proposal and worth explicit discussion?

Reply:
right, hence the quotes :)

the point is to protect the client from a potentially rouge or malicious registry that can cause the client to be very slow by circuit-breaking and failing quickly

Reply:
I think the fact that "server-driven DOS" isn't commonly used term is instructive. At the application layer (Layer 7), it  doesn't make much sense to talk about protecting the client from the server in this way. I don't necessarily see anything about the effect of slow, large, or unexpected response handling that wouldn't also be the case for any other network application. It seems to me like an implementation detail.

Reply:
that's fair and you can choose to title this however you prefer. the higher order bit here is that the proposal is missing a statement about the client (SwiftPM) needing to address this concern by employing techniques such as exponential backoff and circuit breaking - its an implementation detail that is not obvious and should be added.